diff --git a/driver/include/common.h b/driver/include/common.h
index 5cec282..c9f45d2 100644
--- a/driver/include/common.h
+++ b/driver/include/common.h
@@ -16,7 +16,7 @@ typedef struct mp_ptr_t
     void *ptr;
 } mp_ptr_t;
 
-STATIC mp_int_t mp_ptr_get_buffer(mp_obj_t self_in, mp_buffer_info_t *bufinfo, mp_uint_t flags)
+static mp_int_t mp_ptr_get_buffer(mp_obj_t self_in, mp_buffer_info_t *bufinfo, mp_uint_t flags)
 {
     mp_ptr_t *self = MP_OBJ_TO_PTR(self_in);
 
@@ -34,7 +34,7 @@ STATIC mp_int_t mp_ptr_get_buffer(mp_obj_t self_in, mp_buffer_info_t *bufinfo, m
 #define PTR_OBJ(ptr_global) ptr_global ## _obj
 
 #define DEFINE_PTR_OBJ_TYPE(ptr_obj_type, ptr_type_qstr)\
-STATIC MP_DEFINE_CONST_OBJ_TYPE(\
+static MP_DEFINE_CONST_OBJ_TYPE(\
     ptr_obj_type,\
     ptr_type_qstr,\
     MP_TYPE_FLAG_NONE,\
@@ -43,7 +43,7 @@ STATIC MP_DEFINE_CONST_OBJ_TYPE(\
 
 #define DEFINE_PTR_OBJ(ptr_global)\
 DEFINE_PTR_OBJ_TYPE(ptr_global ## _type, MP_QSTR_ ## ptr_global);\
-STATIC const mp_ptr_t PTR_OBJ(ptr_global) = {\
+static const mp_ptr_t PTR_OBJ(ptr_global) = {\
     { &ptr_global ## _type },\
     &ptr_global\
 }
diff --git a/lib/fs_driver.py b/lib/fs_driver.py
index bac84b3..0e0a989 100644
--- a/lib/fs_driver.py
+++ b/lib/fs_driver.py
@@ -3,7 +3,7 @@ Original author: mhepp(https://forum.lvgl.io/u/mhepp/summary)
 '''
 
 import lvgl as lv
-import ustruct as struct
+import struct as struct
 
 def fs_open_cb(drv, path, mode):
 
diff --git a/lv_conf.h b/lv_conf.h
index 2e0d08d..246ec8c 100644
--- a/lv_conf.h
+++ b/lv_conf.h
@@ -704,7 +704,7 @@ extern void mp_lv_deinit_gc();
 #define LV_USE_LIBPNG 0
 
 /*BMP decoder library*/
-#define LV_USE_BMP 0
+#define LV_USE_BMP 1
 
 /* JPG + split JPG decoder library.
  * Split JPG is a custom format optimized for embedded systems. */
