Index: micropython-origin/ports/esp32/boards/make-pins.py
===================================================================
--- micropython-origin.orig/ports/esp32/boards/make-pins.py
+++ micropython-origin/ports/esp32/boards/make-pins.py
@@ -8,7 +8,7 @@ import boardgen
 
 
 # Pins start at zero, and the highest pin index on any ESP32* chip is 48.
-NUM_GPIOS = 49
+NUM_GPIOS = 55
 
 
 class Esp32Pin(boardgen.Pin):
Index: micropython-origin/ports/esp32/machine_pin.h
===================================================================
--- micropython-origin.orig/ports/esp32/machine_pin.h
+++ micropython-origin/ports/esp32/machine_pin.h
@@ -174,6 +174,62 @@
 #define MICROPY_HW_ENABLE_GPIO48 (1)
 #endif
 
+#elif CONFIG_IDF_TARGET_ESP32P4
+#define MICROPY_HW_ENABLE_GPIO0 (1)
+#define MICROPY_HW_ENABLE_GPIO1 (1)
+#define MICROPY_HW_ENABLE_GPIO2 (1)
+#define MICROPY_HW_ENABLE_GPIO3 (1)
+#define MICROPY_HW_ENABLE_GPIO4 (1)
+#define MICROPY_HW_ENABLE_GPIO5 (1)
+#define MICROPY_HW_ENABLE_GPIO6 (1)
+#define MICROPY_HW_ENABLE_GPIO7 (1)
+#define MICROPY_HW_ENABLE_GPIO8 (1)
+#define MICROPY_HW_ENABLE_GPIO9 (1)
+#define MICROPY_HW_ENABLE_GPIO10 (1)
+#define MICROPY_HW_ENABLE_GPIO11 (1)
+#define MICROPY_HW_ENABLE_GPIO12 (1)
+#define MICROPY_HW_ENABLE_GPIO13 (1)
+#define MICROPY_HW_ENABLE_GPIO14 (1)
+#define MICROPY_HW_ENABLE_GPIO15 (1)
+#define MICROPY_HW_ENABLE_GPIO16 (1)
+#define MICROPY_HW_ENABLE_GPIO17 (1)
+#define MICROPY_HW_ENABLE_GPIO18 (1)
+#define MICROPY_HW_ENABLE_GPIO19 (1)
+#define MICROPY_HW_ENABLE_GPIO20 (1)
+#define MICROPY_HW_ENABLE_GPIO21 (1)
+#define MICROPY_HW_ENABLE_GPIO22 (1)
+#define MICROPY_HW_ENABLE_GPIO23 (1)
+#define MICROPY_HW_ENABLE_GPIO24 (0)
+#define MICROPY_HW_ENABLE_GPIO25 (0)
+#define MICROPY_HW_ENABLE_GPIO26 (0)
+#define MICROPY_HW_ENABLE_GPIO27 (0)
+#define MICROPY_HW_ENABLE_GPIO28 (1)
+#define MICROPY_HW_ENABLE_GPIO29 (1)
+#define MICROPY_HW_ENABLE_GPIO30 (1)
+#define MICROPY_HW_ENABLE_GPIO31 (1)
+#define MICROPY_HW_ENABLE_GPIO32 (1)
+#define MICROPY_HW_ENABLE_GPIO33 (1)
+#define MICROPY_HW_ENABLE_GPIO34 (1)
+#define MICROPY_HW_ENABLE_GPIO35 (1)
+#define MICROPY_HW_ENABLE_GPIO36 (1)
+#define MICROPY_HW_ENABLE_GPIO37 (1)
+#define MICROPY_HW_ENABLE_GPIO38 (1)
+#define MICROPY_HW_ENABLE_GPIO39 (1)
+#define MICROPY_HW_ENABLE_GPIO40 (1)
+#define MICROPY_HW_ENABLE_GPIO41 (1)
+#define MICROPY_HW_ENABLE_GPIO42 (1)
+#define MICROPY_HW_ENABLE_GPIO43 (1)
+#define MICROPY_HW_ENABLE_GPIO44 (1)
+#define MICROPY_HW_ENABLE_GPIO45 (1)
+#define MICROPY_HW_ENABLE_GPIO46 (1)
+#define MICROPY_HW_ENABLE_GPIO47 (1)
+#define MICROPY_HW_ENABLE_GPIO48 (1)
+#define MICROPY_HW_ENABLE_GPIO49 (1)
+#define MICROPY_HW_ENABLE_GPIO50 (1)
+#define MICROPY_HW_ENABLE_GPIO51 (1)
+#define MICROPY_HW_ENABLE_GPIO52 (1)
+#define MICROPY_HW_ENABLE_GPIO53 (1)
+#define MICROPY_HW_ENABLE_GPIO54 (1)
 #endif
 
 typedef struct _machine_pin_irq_obj_t {
