Index: micropython/ports/esp32/machine_touchpad.c
===================================================================
--- micropython.orig/ports/esp32/machine_touchpad.c
+++ micropython/ports/esp32/machine_touchpad.c
@@ -29,7 +29,7 @@
 #include "modmachine.h"
 #include "driver/gpio.h"
 
-#if SOC_TOUCH_SENSOR_SUPPORTED
+#if SOC_TOUCH_SENSOR_SUPPORTED && !CONFIG_IDF_TARGET_ESP32P4
 
 #if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 3, 0)
 #if SOC_TOUCH_VERSION_1
Index: micropython/ports/esp32/modmachine.c
===================================================================
--- micropython.orig/ports/esp32/modmachine.c
+++ micropython/ports/esp32/modmachine.c
@@ -45,7 +45,7 @@
 #define MICROPY_PY_MACHINE_SDCARD_ENTRY
 #endif
 
-#if SOC_TOUCH_SENSOR_SUPPORTED
+#if SOC_TOUCH_SENSOR_SUPPORTED && !CONFIG_IDF_TARGET_ESP32P4
 #define MICROPY_PY_MACHINE_TOUCH_PAD_ENTRY { MP_ROM_QSTR(MP_QSTR_TouchPad), MP_ROM_PTR(&machine_touchpad_type) },
 #else
 #define MICROPY_PY_MACHINE_TOUCH_PAD_ENTRY
