Index: micropython/ports/esp32/panichandler.c
===================================================================
--- micropython.orig/ports/esp32/panichandler.c
+++ micropython/ports/esp32/panichandler.c
@@ -42,11 +42,19 @@
 #endif
 
 void __real_esp_panic_handler(void *);
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 4, 2)
+void esp_panic_handler_enable_rtc_wdt(uint32_t timeout_ms);
+#else
 void esp_panic_handler_reconfigure_wdts(uint32_t timeout_ms);
+#endif
 void panic_print_str(const char *str);
 
 void MICROPY_WRAP_PANICHANDLER_FUN(__wrap_esp_panic_handler)(void *info) {
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 4, 2)
+    esp_panic_handler_enable_rtc_wdt(1000);
+#else
     esp_panic_handler_reconfigure_wdts(1000);
+#endif
 
     const static char *msg = MICROPY_WRAP_PANICHANDLER_STR(
         "\r\n"
