Index: micropython/ports/esp32/usb.c
===================================================================
--- micropython.orig/ports/esp32/usb.c
+++ micropython/ports/esp32/usb.c
@@ -44,7 +44,13 @@ void usb_phy_init(void) {
     static const usb_phy_config_t phy_conf = {
         .controller = USB_PHY_CTRL_OTG,
         .otg_mode = USB_OTG_MODE_DEVICE,
+#if CONFIG_IDF_TARGET_ESP32P4 && MICROPY_HW_USB_HS
+        .target = USB_PHY_TARGET_UTMI,
+        .otg_speed = USB_PHY_SPEED_HIGH,
+#else
         .target = USB_PHY_TARGET_INT,
+        .otg_speed = USB_PHY_SPEED_FULL,
+#endif
     };
 
     // Init ESP USB Phy
