Index: M5Unified/src/M5Unified.cpp
===================================================================
--- M5Unified.orig/src/M5Unified.cpp
+++ M5Unified/src/M5Unified.cpp
@@ -67,6 +67,7 @@ static constexpr const uint8_t _pin_tabl
 { board_t::board_M5AirQ       , GPIO_NUM_12,GPIO_NUM_11 , GPIO_NUM_15,GPIO_NUM_13 },
 { board_t::board_M5Cardputer  , 255        ,255         , GPIO_NUM_1 ,GPIO_NUM_2  },
 { board_t::board_M5VAMeter    , GPIO_NUM_6 ,GPIO_NUM_5  , GPIO_NUM_9 ,GPIO_NUM_8  },
+{ board_t::board_M5AtomS3R    , GPIO_NUM_0 ,GPIO_NUM_45 , GPIO_NUM_1 ,GPIO_NUM_2  }, // AtomS3R
 { board_t::board_unknown      , GPIO_NUM_39,GPIO_NUM_38 , GPIO_NUM_1 ,GPIO_NUM_2  }, // AtomS3,AtomS3Lite,AtomS3U
 #elif defined (CONFIG_IDF_TARGET_ESP32C3)
 { board_t::board_unknown      , 255        ,255         , GPIO_NUM_0 ,GPIO_NUM_1  },
@@ -721,6 +722,7 @@ for (int i = 0; i < 0x50; ++i)
     case board_t::board_M5AtomS3:
     case board_t::board_M5AtomS3Lite:
     case board_t::board_M5AtomS3U:
+    case board_t::board_M5AtomS3R:
       m5gfx::pinMode(GPIO_NUM_41, m5gfx::pin_mode_t::input);
       break;
 
@@ -894,6 +896,7 @@ for (int i = 0; i < 0x50; ++i)
 
       case board_t::board_M5AtomS3:
       case board_t::board_M5AtomS3Lite:
+      case board_t::board_M5AtomS3R:
         if (cfg.external_speaker.atomic_spk && (Display.getBoard() != board_t::board_M5AtomDisplay))
         { // for ATOMIC SPK
           m5gfx::pinMode(GPIO_NUM_6, m5gfx::pin_mode_t::input_pulldown); // MOSI
@@ -1301,6 +1304,7 @@ for (int i = 0; i < 0x50; ++i)
     case board_t::board_M5AtomS3:
     case board_t::board_M5AtomS3Lite:
     case board_t::board_M5AtomS3U:
+    case board_t::board_M5AtomS3R:
       use_rawstate_bits = 0b00001;
       btn_rawstate_bits = (!m5gfx::gpio_in(GPIO_NUM_41)) & 1;
       break;
Index: M5Unified/src/M5Unified.hpp
===================================================================
--- M5Unified.orig/src/M5Unified.hpp
+++ M5Unified/src/M5Unified.hpp
@@ -340,7 +340,7 @@ namespace m5
 #if defined ( __M5GFX_M5ATOMDISPLAY__ )
 #if !defined (CONFIG_IDF_TARGET) || defined (CONFIG_IDF_TARGET_ESP32) || defined (CONFIG_IDF_TARGET_ESP32S3)
       if (cfg.external_display.atom_display) {
-        if (_board == board_t::board_M5Atom || _board == board_t::board_M5AtomPsram || _board == board_t::board_M5AtomS3 || _board == board_t::board_M5AtomS3Lite)
+        if (_board == board_t::board_M5Atom || _board == board_t::board_M5AtomPsram || _board == board_t::board_M5AtomS3 || _board == board_t::board_M5AtomS3Lite || _board == board_t::board_M5AtomS3R)
         {
           M5AtomDisplay dsp(cfg.atom_display);
           if (dsp.init_without_reset()) {
