Index: M5Unified/src/M5Unified.cpp
===================================================================
--- M5Unified.orig/src/M5Unified.cpp
+++ M5Unified/src/M5Unified.cpp
@@ -1418,6 +1418,24 @@ static constexpr const uint8_t _pin_tabl
         ioexp->begin();
         _io_expander[i].reset(ioexp);
       }
+        // sx1262 reset(P7)
+        _io_expander[0]->setDirection(7, true);
+        _io_expander[0]->setPullMode(7, false);
+        _io_expander[0]->setHighImpedance(7, false);
+        _io_expander[0]->digitalWrite(7, false);
+        delay(10);
+        _io_expander[0]->digitalWrite(7, true);
+        delay(10);
+        // LAN EN
+        _io_expander[0]->setDirection(5, true);
+        _io_expander[0]->setPullMode(5, false);
+        _io_expander[0]->setHighImpedance(5, false);
+        _io_expander[0]->digitalWrite(5, true);
+        // SW EN
+        _io_expander[0]->setDirection(6, true);
+        _io_expander[0]->setPullMode(6, false);
+        _io_expander[0]->setHighImpedance(6, false);
+        _io_expander[0]->digitalWrite(6, true);
       break;
 #elif defined (CONFIG_IDF_TARGET_ESP32S3)
     case board_t::board_M5StampPLC:
