diff --git a/src/utility/Power_Class.cpp b/src/utility/Power_Class.cpp
index cd55d3b..1bcddb0 100644
--- a/src/utility/Power_Class.cpp
+++ b/src/utility/Power_Class.cpp
@@ -1455,7 +1455,7 @@ namespace m5
       case board_t::board_M5PowerHub:
         uint8_t buf[2];
         if(M5.In_I2C.readRegister(powerhub_i2c_addr, 0x32, buf, sizeof(buf), i2c_freq))
-          return (int16_t)(buf[1] << 8) | buf[0];
+          return -(int16_t)((buf[1] << 8) | buf[0]);
         return 0;
 #endif
       default:
