diff --git a/src/utility/Power_Class.cpp b/src/utility/Power_Class.cpp
index cd55d3b..e96fb47 100644
--- a/src/utility/Power_Class.cpp
+++ b/src/utility/Power_Class.cpp
@@ -1313,7 +1313,7 @@ namespace m5
 
 #if defined (CONFIG_IDF_TARGET_ESP32S3)
       case board_t::board_M5PowerHub:
-        mv = getBatteryVoltage();
+        mv = getBatteryVoltage() / 2;
         break;
 #endif
       default:
