diff --git a/src/utility/Power_Class.cpp b/src/utility/Power_Class.cpp
index cd55d3b..cb5b044 100644
--- a/src/utility/Power_Class.cpp
+++ b/src/utility/Power_Class.cpp
@@ -210,6 +210,18 @@ namespace m5
     case board_t::board_M5PowerHub:
       M5.In_I2C.writeRegister8(powerhub_i2c_addr, 0x05, 1, i2c_freq); // Enabel VAMeter
       break;
+    
+    case board_t::board_M5StampPLC:
+      Ina226.begin();
+      INA226_Class::config_t cfg;
+      cfg.sampling_rate = INA226_Class::Sampling::Rate16;
+      cfg.bus_conversion_time = INA226_Class::ConversionTime::US_1100;
+      cfg.shunt_conversion_time = INA226_Class::ConversionTime::US_1100;
+      cfg.mode = INA226_Class::Mode::ShuntAndBus;
+      cfg.shunt_res = 0.01f;
+      cfg.max_expected_current = 8.192f;
+      Ina226.config(cfg);
+      break;
     }
 
 #elif !defined (CONFIG_IDF_TARGET) || defined (CONFIG_IDF_TARGET_ESP32)
@@ -1550,11 +1562,11 @@ namespace m5
     }
   }
 
-  int16_t Power_Class::_readExtValue(ext_port_mask_t port_mask, int reg_offset)
+  int16_t Power_Class::_readExtValue(ext_port_mask_t port_mask, bool is_voltage)
   {
 #if defined(M5UNIFIED_PC_BUILD)
       (void)port_mask;
-      (void)reg_offset;
+      (void)is_voltage;
 #else
     switch (M5.getBoard()) {
     #if defined(CONFIG_IDF_TARGET_ESP32S3)
@@ -1574,14 +1586,23 @@ namespace m5
         uint8_t buf[2];
         for (const auto& pr : port_regs) {
           if (port_mask & pr.mask) {
-            if (M5.In_I2C.readRegister(powerhub_i2c_addr, pr.reg + reg_offset, buf, sizeof(buf), i2c_freq)) {
+            if (M5.In_I2C.readRegister(powerhub_i2c_addr, pr.reg + (is_voltage ? 0 : 2), buf, sizeof(buf), i2c_freq)) {
               return (int16_t)((buf[1] << 8) | buf[0]);
             }
               return 0;
           }
         }
         return 0;
-      } break;
+      }
+
+      case board_t::board_M5StampPLC:
+        if (port_mask & (ext_port_mask_t::ext_PWR485 | ext_port_mask_t::ext_PWRCAN)) {
+          if (is_voltage)
+            return Ina226.getBusVoltage() * 1000;
+          else
+            return 0;
+        }
+        return 0;
     #endif
       default:
         return 0;
@@ -1591,12 +1612,12 @@ namespace m5
 
   int16_t Power_Class::getExtVoltage(ext_port_mask_t port_mask)
   {
-    return _readExtValue(port_mask, 0);
+    return _readExtValue(port_mask, true);
   }
 
   int16_t Power_Class::getExtCurrent(ext_port_mask_t port_mask)
   {
-    return _readExtValue(port_mask, 2);
+    return _readExtValue(port_mask, false);
   }
 
   uint8_t Power_Class::getKeyState(void)
diff --git a/src/utility/Power_Class.hpp b/src/utility/Power_Class.hpp
index 56e485b..36e6b63 100644
--- a/src/utility/Power_Class.hpp
+++ b/src/utility/Power_Class.hpp
@@ -205,6 +205,7 @@ namespace m5
 #if defined (CONFIG_IDF_TARGET_ESP32S3)
 
     AXP2101_Class Axp2101;
+    INA226_Class Ina226 = { 0x40 };
 
 #elif defined (CONFIG_IDF_TARGET_ESP32C3)
 #elif defined (CONFIG_IDF_TARGET_ESP32C6)
@@ -229,7 +230,7 @@ namespace m5
     std::int32_t _getBatteryAdcRaw(void);
     void _powerOff(bool withTimer);
     void _timerSleep(void);
-    int16_t _readExtValue(ext_port_mask_t port_mask, int reg_offset);
+    int16_t _readExtValue(ext_port_mask_t port_mask, bool is_voltage);
 
     float _adc_ratio = 0;
     std::uint8_t _wakeupPin = 255;
