Index: M5Unified/src/M5Unified.cpp
===================================================================
--- M5Unified.orig/src/M5Unified.cpp
+++ M5Unified/src/M5Unified.cpp
@@ -10,12 +10,25 @@
 #include <soc/efuse_reg.h>
 #include <soc/gpio_periph.h>
 
-#if !defined (CONFIG_IDF_TARGET) || defined (CONFIG_IDF_TARGET_ESP32)
- #if __has_include (<driver/touch_sens.h>)
- #include <driver/touch_sens.h>
- #elif __has_include (<driver/touch_sensor.h>)
- #include <driver/touch_sensor.h>
- #endif
+#if SOC_TOUCH_SENSOR_SUPPORTED
+#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 3, 0)
+#if SOC_TOUCH_VERSION_1
+#define SOC_TOUCH_SENSOR_VERSION (1)
+#elif SOC_TOUCH_VERSION_2
+#define SOC_TOUCH_SENSOR_VERSION (2)
+#endif
+#endif
+
+#if SOC_TOUCH_SENSOR_VERSION == 1 // ESP32 only
+#include "driver/touch_pad.h"
+#elif SOC_TOUCH_SENSOR_VERSION == 2 // most ESP32
+#include "driver/touch_sensor.h"
+#elif SOC_TOUCH_SENSOR_VERSION == 3 // At present, it can only be used on ESP32P4.
+#include "driver/touch_sens.h"
+#include "soc/touch_sensor_channel.h"
+#else
+#error "Unknown touch hardware version"
+#endif
 #endif
 
 #if __has_include (<driver/i2s_type.h>)
@@ -1215,5 +1228,5 @@ static bool _read_touch_pad(uint32_t* re
   static bool _read_touch_pad(uint32_t* results, const int* channel, const size_t channel_count)
   {
     for (size_t i = 0; i < channel_count; ++i) { results[i] = 0; }
-#if defined ( TOUCH_SENSOR_DEFAULT_FILTER_CONFIG )
+#if SOC_TOUCH_SENSOR_VERSION == 3
     if (channel_count > TOUCH_TOTAL_CHAN_NUM) { return false; }
