Index: M5Unified/src/utility/Power_Class.cpp
===================================================================
--- M5Unified.orig/src/utility/Power_Class.cpp
+++ M5Unified/src/utility/Power_Class.cpp
@@ -347,6 +347,10 @@ namespace m5
           M5_LOGE("M5IOE1 CHG_PROG pull state could not be released.");
         }
         ioe1.setDirection(M5IOE1_Class::gpio3, false);
+        // Turn on SD card power (IOE1 G6).
+        ioe1.setHighImpedance(M5IOE1_Class::gpio6, false);
+        ioe1.setDirection(M5IOE1_Class::gpio6, true);
+        ioe1.digitalWrite(M5IOE1_Class::gpio6, true);
       }
       M5pm1.setBatteryCharge(true);
       M5pm1.setDCDCOutput(true);
