Index: M5GFX/src/lgfx/v1/platforms/esp32/common.cpp
===================================================================
--- M5GFX.orig/src/lgfx/v1/platforms/esp32/common.cpp
+++ M5GFX/src/lgfx/v1/platforms/esp32/common.cpp
@@ -1382,17 +1382,19 @@ namespace lgfx
  #endif
 #elif __has_include(<driver/i2c_master.h>)
       i2c_master_bus_handle_t bus_handle = nullptr;
-      i2c_master_bus_config_t bus_config;
-      memset(&bus_config, 0, sizeof(i2c_master_bus_config_t));
-      bus_config.i2c_port = i2c_port;
-      bus_config.sda_io_num = pin_sda;
-      bus_config.scl_io_num = pin_scl;
-      bus_config.clk_source = I2C_CLK_SRC_DEFAULT;
-      bus_config.glitch_ignore_cnt = 7;
-      bus_config.flags.enable_internal_pullup = true;
-      bus_config.intr_priority = 1;
+      if (i2c_master_get_bus_handle(i2c_port, &bus_handle) != ESP_OK) {
+        i2c_master_bus_config_t bus_config;
+        memset(&bus_config, 0, sizeof(i2c_master_bus_config_t));
+        bus_config.i2c_port = i2c_port;
+        bus_config.sda_io_num = pin_sda;
+        bus_config.scl_io_num = pin_scl;
+        bus_config.clk_source = I2C_CLK_SRC_DEFAULT;
+        bus_config.glitch_ignore_cnt = 7;
+        bus_config.flags.enable_internal_pullup = true;
+        bus_config.intr_priority = 1;
 
-      i2c_new_master_bus(&bus_config, &bus_handle);
+        i2c_new_master_bus(&bus_config, &bus_handle);
+      }
       i2c_context[i2c_port].i2c_bus_handle = bus_handle;
 #else
       i2c_periph_enable(i2c_port);
