diff --git a/lv_conf.h b/lv_conf.h
index 2e0d08d..5d36bca 100644
--- a/lv_conf.h
+++ b/lv_conf.h
@@ -421,27 +421,69 @@ extern void mp_lv_deinit_gc();
 
 /*Montserrat fonts with ASCII range and some symbols using bpp = 4
  *https://fonts.google.com/specimen/Montserrat*/
-#define LV_FONT_MONTSERRAT_8  0
-#define LV_FONT_MONTSERRAT_10 0
-#define LV_FONT_MONTSERRAT_12 0
-#define LV_FONT_MONTSERRAT_14 1
-#define LV_FONT_MONTSERRAT_16 1
-#define LV_FONT_MONTSERRAT_18 0
-#define LV_FONT_MONTSERRAT_20 0
-#define LV_FONT_MONTSERRAT_22 0
-#define LV_FONT_MONTSERRAT_24 1
-#define LV_FONT_MONTSERRAT_26 0
-#define LV_FONT_MONTSERRAT_28 0
-#define LV_FONT_MONTSERRAT_30 0
-#define LV_FONT_MONTSERRAT_32 0
-#define LV_FONT_MONTSERRAT_34 0
-#define LV_FONT_MONTSERRAT_36 0
-#define LV_FONT_MONTSERRAT_38 0
-#define LV_FONT_MONTSERRAT_40 0
-#define LV_FONT_MONTSERRAT_42 0
-#define LV_FONT_MONTSERRAT_44 0
-#define LV_FONT_MONTSERRAT_46 0
-#define LV_FONT_MONTSERRAT_48 0
+#ifndef LV_FONT_MONTSERRAT_8
+    #define LV_FONT_MONTSERRAT_8  0
+#endif
+#ifndef LV_FONT_MONTSERRAT_10
+    #define LV_FONT_MONTSERRAT_10 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_12
+    #define LV_FONT_MONTSERRAT_12 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_14
+    #define LV_FONT_MONTSERRAT_14 1
+#endif
+#ifndef LV_FONT_MONTSERRAT_16
+    #define LV_FONT_MONTSERRAT_16 1
+#endif
+#ifndef LV_FONT_MONTSERRAT_18
+    #define LV_FONT_MONTSERRAT_18 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_20
+    #define LV_FONT_MONTSERRAT_20 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_22
+    #define LV_FONT_MONTSERRAT_22 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_24
+    #define LV_FONT_MONTSERRAT_24 1
+#endif
+#ifndef LV_FONT_MONTSERRAT_26
+    #define LV_FONT_MONTSERRAT_26 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_28
+    #define LV_FONT_MONTSERRAT_28 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_30
+    #define LV_FONT_MONTSERRAT_30 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_32
+    #define LV_FONT_MONTSERRAT_32 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_34
+    #define LV_FONT_MONTSERRAT_34 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_36
+    #define LV_FONT_MONTSERRAT_36 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_38
+    #define LV_FONT_MONTSERRAT_38 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_40
+    #define LV_FONT_MONTSERRAT_40 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_42
+    #define LV_FONT_MONTSERRAT_42 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_44
+    #define LV_FONT_MONTSERRAT_44 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_46
+    #define LV_FONT_MONTSERRAT_46 0
+#endif
+#ifndef LV_FONT_MONTSERRAT_48
+    #define LV_FONT_MONTSERRAT_48 0
+#endif
 
 /*Demonstrate special features*/
 #define LV_FONT_MONTSERRAT_28_COMPRESSED 0  /*bpp = 3*/
